core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowRewriter.scala

```diff
@@ -0,0 +1,424 @@
+package edu.uci.ics.texera.workflow.common.workflow
+
+import com.typesafe.scalalogging.Logger
+import edu.uci.ics.texera.workflow.common.storage.OpResultStorage
+import edu.uci.ics.texera.Utils.objectMapper
+import edu.uci.ics.texera.workflow.common.operators.OperatorDescriptor
+import edu.uci.ics.texera.workflow.common.workflow.WorkflowRewriter.copyOperator
+import edu.uci.ics.texera.workflow.operators.sink.CacheSinkOpDesc
+import edu.uci.ics.texera.workflow.operators.source.cache.CacheSourceOpDesc
+
+import java.util.UUID
+import scala.collection.mutable
+
+case class WorkflowVertex(
+    op: OperatorDescriptor,
+    links: mutable.HashSet[OperatorLink]
+)
+
+object WorkflowRewriter {
+  private def copyOperator(operator: OperatorDescriptor): OperatorDescriptor = {
+    objectMapper.readValue(objectMapper.writeValueAsString(operator), classOf[OperatorDescriptor])
+  }
+}
+
+class WorkflowRewriter(
+    var workflowInfo: WorkflowInfo,
+    var cachedOperatorDescriptors: mutable.HashMap[String, OperatorDescriptor],
+    var cacheSourceOperatorDescriptors: mutable.HashMap[String, CacheSourceOpDesc],
+    var cacheSinkOperatorDescriptors: mutable.HashMap[String, CacheSinkOpDesc],
+    var operatorRecord: mutable.HashMap[String, WorkflowVertex],
+    var opResultStorage: OpResultStorage
+) {
+
+  private val logger = Logger(this.getClass.getName)
+
+  var visitedOpID: mutable.HashSet[String] = new mutable.HashSet[String]()
+
+  private val workflowDAG: WorkflowDAG = if (workflowInfo != null) {
```

Comment:
Can we use `Option` to avoid the presence of `null`? Nullpointer can be pretty hard to debug.

Reply:
Can we simply forces the workflowInfo to be not null?

Reply:
This is a question for @zuozhiw ?